File: backend/schemas/analysis.py
from pydantic import BaseModel, Field, model_validator


class AnalyzeRequest(BaseModel):
    """Input for AI paper analysis: a title plus either an abstract or full PDF text."""

    title: str = Field(..., min_length=1)
    abstract: str | None = Field(default=None, description="Paper abstract")
    pdf_text: str | None = Field(
        default=None, description="Full extracted PDF text, used when no abstract is available"
    )

    @model_validator(mode="after")
    def _require_content(self) -> "AnalyzeRequest":
        if not self.abstract and not self.pdf_text:
            raise ValueError("Provide either 'abstract' or 'pdf_text'")
        return self


class PaperAnalysis(BaseModel):
    """Structured battery-research metadata extracted by the AI analysis pipeline.

    Every field mirrors the requested JSON schema exactly. Descriptive
    fields are nullable (the model reports null rather than guessing when
    the source text doesn't mention something); list fields default to
    empty rather than null for easier consumption by callers.
    """

    title: str | None = None
    authors: list[str] = Field(default_factory=list)
    journal: str | None = None
    battery_system: str | None = Field(
        default=None, description="Battery chemistry class, e.g. Li-ion, Li-metal, Na-ion, solid-state"
    )
    electrolyte: str | None = None
    salt: str | None = None
    solvent: str | None = None
    additive: str | None = None
    cathode: str | None = None
    anode: str | None = None
    separator: str | None = None
    cell_type: str | None = None
    voltage_window: str | None = None
    temperature: str | None = None
    formation_protocol: str | None = None
    cycle_condition: str | None = None
    rate_capability: str | None = None
    main_findings: list[str] = Field(default_factory=list)
    innovation: str | None = None
    advantages: list[str] = Field(default_factory=list)
    limitations: list[str] = Field(default_factory=list)
    future_work: list[str] = Field(default_factory=list)

File: frontend/src/types/analysis.ts
// Mirrors backend/schemas/analysis.py.
export interface AnalyzeRequest {
  title: string
  abstract?: string
  pdf_text?: string
}

export interface PaperAnalysis {
  title: string | null
  authors: string[]
  journal: string | null
  battery_system: string | null
  electrolyte: string | null
  salt: string | null
  solvent: string | null
  additive: string | null
  cathode: string | null
  anode: string | null
  separator: string | null
  cell_type: string | null
  voltage_window: string | null
  temperature: string | null
  formation_protocol: string | null
  cycle_condition: string | null
  rate_capability: string | null
  main_findings: string[]
  innovation: string | null
  advantages: string[]
  limitations: string[]
  future_work: string[]
}
